pkg/engine/lifecycletest/fuzzing/resource.go

```diff
@@ -0,0 +1,374 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuzzing
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/mitchellh/copystructure"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy/providers"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"golang.org/x/exp/maps"
+	"pgregory.net/rapid"
+)
+
+// A ResourceSpec specifies the subset of a resource's state that is relevant to fuzzing snapshot integrity issues.
+// Generally this encompasses enough to identify a resource (URN, ID, and so on) and any dependencies it may have on
+// others.
+type ResourceSpec struct {
+	Project              tokens.PackageName
+	Stack                tokens.QName
+	Type                 tokens.Type
+	Name                 string
+	ID                   resource.ID
+	Custom               bool
+	Delete               bool
+	Protect              bool
+	PendingReplacement   bool
+	RetainOnDelete       bool
+	Provider             string
+	Dependencies         []resource.URN
+	PropertyDependencies map[resource.PropertyKey][]resource.URN
+	DeletedWith          resource.URN
+
+	// A set of tags associated with the resource. These have no bearing on any tests but are included to aid in debugging
+	// and identifying the causes of snapshot integrity issues.
+	Tags map[string]bool
+}
+
+// Adds the given tag to the given ResourceSpec.
+func AddTag[T ~string](r *ResourceSpec, tag T) {
+	if tag == "" {
+		return
+	}
+
+	if r.Tags == nil {
+		r.Tags = map[string]bool{}
+	}
+
+	r.Tags[string(tag)] = true
+}
+
+// Returns the URN of this ResourceSpec.
+func (r *ResourceSpec) URN() resource.URN {
```

Comment:
Nit: I think the go convention is to start the doc comment with the name of the thing being described, for example: `URN returns URN of this ResourceSpec.` (yes this does often lead to duplication 🤷 )